File: Python/decision_tree.py
from sklearn.tree import DecisionTreeClassifier, plot_tree
import pandas as pd
import time
import numpy as np

# Load your data into a pandas dataframe
data = pd.read_csv('C:/Users/Hyrum Hansen/Documents/github/model-benchmarking/data/classification_data.csv')

# Split the data into feature matrix (X) and target vector (y)
X = data[['x1', 'x2', 'x3']]
y = data['y']

n = 500 # Number of times to build the decision tree
train_times = [0] * n # List to store training times for each iteration
for i in range(n):

    start_time = time.time() # Start timing
    clf = DecisionTreeClassifier()
    clf.fit(X, y)
    train_time = time.time() - start_time # Calculate training time

    train_times[i] = train_time

# Calculate statistics
total_time = sum(train_times)
median_time = np.median(train_times)
mean_time = np.mean(train_times)
min_time = min(train_times)
max_time = max(train_times)

# Print results
print(f"Total time: {total_time:.4f} seconds")
print(f"Median time: {median_time:.4f} seconds")
print(f"Mean time: {mean_time:.4f} seconds")
print(f"Minimum time: {min_time:.4f} seconds")
print(f"Maximum time: {max_time:.4f} seconds")

File: Julia/decision_tree.jl
using DecisionTree, CSV, DataFrames, MLBase

# Load data from CSV file
data = CSV.read("C:/Users/Hyrum Hansen/Documents/github/model-benchmarking/data/classification_data.csv", DataFrame)

features = Matrix(select(data, Not([:Column1, :y])))
labels = data[!, :y]

# Initialize an empty vector to store the fit times
fit_times = Float64[]

# Loop over n iterations and fit the decision tree model
for i in 1:500

    t0 = time()
    model = DecisionTreeClassifier(max_depth=2)
    fit!(model, features, labels) 
    t1 = time()

    fit_time = t1 - t0
    push!(fit_times, fit_time)
    
end

# Calculate the total time, median time, minimum time, maximum time, and average time
total_time = sum(fit_times)
median_time = median(fit_times)
min_time = minimum(fit_times)
max_time = maximum(fit_times)
avg_time = mean(fit_times)

# Print the results to the console
println("Total time: ", total_time)
println("Median time: ", median_time)
println("Minimum time: ", min_time)
println("Maximum time: ", max_time)
println("Average time: ", avg_time)